node/pkg/near/watcher.go

```diff
@@ -0,0 +1,395 @@
+					zap.Binary("payload", observation.Payload),
+					zap.Uint8("consistency_level", observation.ConsistencyLevel),
+				)
+
+				e.msgChan <- observation
+			}
+		}
+	}
+
+	return nil
+}
+
+func inspectBody(e *Watcher, logger *zap.Logger, block uint64, body gjson.Result) error {
+	logger.Info("inspectBody", zap.Uint64("block", block))
+
+	result := body.Get("result.chunks.#.chunk_hash")
+	if !result.Exists() {
+		return nil
+	}
+
+	v := body.Get("result.header.timestamp")
+	if !v.Exists() {
+		logger.Info("result.header.timestamp")
+		return nil
+	}
+	ts := uint64(v.Uint()) / 1000000000
+
+	for _, name := range result.Array() {
+		chunk, err := getChunk(e, name.String())
+		if err != nil {
+			return err
+		}
+
+		txns := gjson.ParseBytes(chunk).Get("result.transactions")
+		if !txns.Exists() {
+			continue
+		}
+		for _, r := range txns.Array() {
+			hash := r.Get("hash")
+			receiver_id := r.Get("receiver_id")
+			if !hash.Exists() || !receiver_id.Exists() {
+				continue
+			}
+
+			err = inspectStatus(e, logger, hash.String(), receiver_id.String(), ts)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func (e *Watcher) Run(ctx context.Context) error {
+	// an odd thing to broadcast...
+	p2p.DefaultRegistry.SetNetworkStats(vaa.ChainIDNear, &gossipv1.Heartbeat_Network{
+		ContractAddress: e.wormholeContract,
+	})
+
+	logger := supervisor.Logger(ctx)
+	errC := make(chan error)
+
+	logger.Info("Near watcher connecting to RPC node ", zap.String("url", e.nearRPC))
+
+	go func() {
+		timer := time.NewTicker(time.Second * 1)
+		defer timer.Stop()
+
+		if e.next_round == 0 {
+			finalBody, err := getFinalBlock(e)
+			if err != nil {
+				logger.Error("StatusAfterBlock", zap.Error(err))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDNear, 1)
+				errC <- err
+				return
+			}
+			e.next_round = gjson.ParseBytes(finalBody).Get("result.chunks.0.height_created").Uint()
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case r := <-e.obsvReqC:
+				if vaa.ChainID(r.ChainId) != vaa.ChainIDNear {
+					panic("invalid chain ID")
+				}
+
+				txHash := base58.Encode(r.TxHash)
+
+				logger.Info("Received obsv request", zap.String("tx_hash", txHash))
+
+				err := inspectStatus(e, logger, txHash, e.wormholeContract, 0)
+				if err != nil {
+					logger.Error(fmt.Sprintf("near obsvReqC: %s", err.Error()))
+				}
+
+			case <-timer.C:
+				finalBody, err := getFinalBlock(e)
+				if err != nil {
+					logger.Error(fmt.Sprintf("nearClient.Status: %s", err.Error()))
+
+					p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDNear, 1)
+					errC <- err
+					return
+				} else {
+					parsedFinalBody := gjson.ParseBytes(finalBody)
+					lastBlock := parsedFinalBody.Get("result.chunks.0.height_created").Uint()
+
+					logger.Info("lastBlock", zap.Uint64("lastBlock", lastBlock), zap.Uint64("next_round", e.next_round))
+
+					if lastBlock < e.next_round {
+						logger.Info("Went backwards... ")
```

Comment:
Make this an error so someone might notice it's happening?

Reply:
agreed.  I wonder if we will ever see this in public